File: VB/DXApplication5/AgendViewComponent/ListViewMenuBuilder.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.XtraScheduler
Imports DevExpress.XtraGrid.Menu
Imports System.Drawing
Imports System.Reflection

Namespace ListViewComponent
	Public NotInheritable Class ListViewMenuBuilder
		Public Shared Property Scheduler() As SchedulerControl
		Public Shared Property CurrentAppointment() As Appointment

		Public Shared Property ViewControl() As ListViewControl

		Private Sub New()
		End Sub
		Public Shared Sub GenerateContextMenu(ByVal viewControl As ListViewControl, ByVal contextMenu As GridViewMenu, ByVal control As SchedulerControl, ByVal apt As Appointment)
			Scheduler = control
			CurrentAppointment = apt
			ViewControl = viewControl
			If apt IsNot Nothing Then
				contextMenu.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Open appointment", AddressOf OnOpenCurrentAppointment))
                contextMenu.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Delete appointment", AddressOf OnDeleteCurrentAppointment, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("Delete.png"))))
			End If


			Dim switchView As New DevExpress.Utils.Menu.DXSubMenuItem("Change view to")
			switchView.BeginGroup = True
			contextMenu.Items.Add(switchView)

			If Scheduler.DayView.Enabled Then
                switchView.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Day View", AddressOf OnSwitchToDayView, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("DayView.png"))))
			End If
			If Scheduler.WorkWeekView.Enabled Then
                switchView.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Work Week View", AddressOf OnSwitchToWorkWeekView, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("WorkWeekView.png"))))
			End If
			If Scheduler.WeekView.Enabled Then
                switchView.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Week View", AddressOf OnSwitchToWeekView, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("WeekView.png"))))
			End If
			If Scheduler.MonthView.Enabled Then
                switchView.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Month View", AddressOf OnSwitchToMonthView, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("MonthView.png"))))
			End If
			If Scheduler.TimelineView.Enabled Then
                switchView.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Timeline View", AddressOf OnSwitchToTimelineView, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("TimelineView.png"))))
			End If
			If Scheduler.GanttView.Enabled Then
                switchView.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Gantt View", AddressOf OnSwitchToGanttView, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("GanttView.png"))))
			End If


		End Sub

		Public Shared Sub OnOpenCurrentAppointment(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.ShowEditAppointmentForm(CurrentAppointment, CurrentAppointment.IsRecurring)
			ViewControl.ChangeAppointment(CurrentAppointment)

		End Sub

		Public Shared Sub OnDeleteCurrentAppointment(ByVal sender As Object, ByVal e As EventArgs)
			ViewControl.DeleteAppointment(CurrentAppointment)
			Scheduler.DeleteAppointment(CurrentAppointment)

		End Sub

		Public Shared Sub OnSwitchToDayView(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.ActiveViewType = SchedulerViewType.Day
			ListViewHelper.SwitchToNormalView()
		End Sub

		Public Shared Sub OnSwitchToWorkWeekView(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.ActiveViewType = SchedulerViewType.WorkWeek
			ListViewHelper.SwitchToNormalView()
		End Sub

		Public Shared Sub OnSwitchToWeekView(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.ActiveViewType = SchedulerViewType.Week
			ListViewHelper.SwitchToNormalView()
		End Sub

		Public Shared Sub OnSwitchToMonthView(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.ActiveViewType = SchedulerViewType.Month
			ListViewHelper.SwitchToNormalView()
		End Sub

		Public Shared Sub OnSwitchToTimelineView(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.ActiveViewType = SchedulerViewType.Timeline
			ListViewHelper.SwitchToNormalView()
		End Sub

		Public Shared Sub OnSwitchToGanttView(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.ActiveViewType = SchedulerViewType.Gantt
			ListViewHelper.SwitchToNormalView()
		End Sub


	End Class
End Namespace

File: CS/DXApplication5/AgendViewComponent/ListViewMenuBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DevExpress.XtraScheduler;
using DevExpress.XtraGrid.Menu;
using System.Drawing;
using System.Reflection;

namespace ListViewComponent {
    public static class ListViewMenuBuilder {
        public static SchedulerControl Scheduler { get; set; }
        public static Appointment CurrentAppointment { get; set; }

        public static ListViewControl ViewControl {
            get;
            set;
        }

        public static void GenerateContextMenu(ListViewControl viewControl, GridViewMenu contextMenu, SchedulerControl control, Appointment apt) {
            Scheduler = control;
            CurrentAppointment = apt;
            ViewControl = viewControl;
            if(apt != null) {
                contextMenu.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Open appointment", OnOpenCurrentAppointment));
                contextMenu.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Delete appointment", OnDeleteCurrentAppointment, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("ListViewComponent.Resources.Delete.png"))));                
            }
            

            DevExpress.Utils.Menu.DXSubMenuItem switchView = new DevExpress.Utils.Menu.DXSubMenuItem("Change view to");
            switchView.BeginGroup = true;
            contextMenu.Items.Add(switchView);

            if (Scheduler.DayView.Enabled) switchView.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Day View", OnSwitchToDayView, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("ListViewComponent.Resources.DayView.png"))));
            if (Scheduler.WorkWeekView.Enabled) switchView.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Work Week View", OnSwitchToWorkWeekView, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("ListViewComponent.Resources.WorkWeekView.png"))));
            if (Scheduler.WeekView.Enabled) switchView.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Week View", OnSwitchToWeekView, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("ListViewComponent.Resources.WeekView.png"))));
            if (Scheduler.MonthView.Enabled) switchView.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Month View", OnSwitchToMonthView, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("ListViewComponent.Resources.MonthView.png"))));
            if (Scheduler.TimelineView.Enabled) switchView.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Timeline View", OnSwitchToTimelineView, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("ListViewComponent.Resources.TimelineView.png"))));
            if (Scheduler.GanttView.Enabled) switchView.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Gantt View", OnSwitchToGanttView, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("ListViewComponent.Resources.GanttView.png"))));

            
        }

        public static void OnOpenCurrentAppointment(object sender, EventArgs e) {
            Scheduler.ShowEditAppointmentForm(CurrentAppointment, CurrentAppointment.IsRecurring);
            ViewControl.ChangeAppointment(CurrentAppointment);
            
        }

        public static void OnDeleteCurrentAppointment(object sender, EventArgs e) {
            ViewControl.DeleteAppointment(CurrentAppointment);
            Scheduler.DeleteAppointment(CurrentAppointment);
      
        }
     
        public static void OnSwitchToDayView(object sender, EventArgs e) {
            Scheduler.ActiveViewType = SchedulerViewType.Day;
            ListViewHelper.SwitchToNormalView();
        }

        public static void OnSwitchToWorkWeekView(object sender, EventArgs e) {
            Scheduler.ActiveViewType = SchedulerViewType.WorkWeek;
            ListViewHelper.SwitchToNormalView();
        }

        public static void OnSwitchToWeekView(object sender, EventArgs e) {
            Scheduler.ActiveViewType = SchedulerViewType.Week;
            ListViewHelper.SwitchToNormalView();
        }

        public static void OnSwitchToMonthView(object sender, EventArgs e) {
            Scheduler.ActiveViewType = SchedulerViewType.Month;
            ListViewHelper.SwitchToNormalView();
        }

        public static void OnSwitchToTimelineView(object sender, EventArgs e) {
            Scheduler.ActiveViewType = SchedulerViewType.Timeline;
            ListViewHelper.SwitchToNormalView();
        }

        public static void OnSwitchToGanttView(object sender, EventArgs e) {
            Scheduler.ActiveViewType = SchedulerViewType.Gantt;
            ListViewHelper.SwitchToNormalView();
        }

        
    }
}
